pkg/logql/blocker.go

```diff
@@ -0,0 +1,94 @@
+package logql
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/regexp"
+
+	logutil "github.com/grafana/loki/pkg/util/log"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type queryBlocker struct {
+	ctx    context.Context
+	q      *query
+	logger log.Logger
+}
+
+func newQueryBlocker(ctx context.Context, q *query) *queryBlocker {
+	return &queryBlocker{
+		ctx:    ctx,
+		q:      q,
+		logger: logutil.WithContext(ctx, q.logger),
+	}
+}
+
+func (qb *queryBlocker) isBlocked(tenant string) bool {
+	patterns := qb.q.limits.BlockedQueries(tenant)
+	if len(patterns) <= 0 {
+		return false
+	}
+
+	typ, err := QueryType(qb.q.params.Query())
+	if err != nil {
+		typ = "unknown"
+	}
+
+	logger := log.With(qb.logger, "user", tenant, "type", typ)
+
+	query := qb.q.params.Query()
+	for _, p := range patterns {
+
+		// if no pattern is given, assume we want to match all queries
+		if p.Pattern == "" {
+			p.Pattern = ".*"
+			p.Regex = true
+		}
```

Comment:
This is also mention worthy in the docs WDYT?

Reply:
I don't think so; it's an implementation detail.

This example points out that blocking can be done purely on types.

```yaml
...
      # block all metric queries
      - types: metric
...
```